### PR TITLE
Refactored bpda.py to support multi-tensor input

### PR DESCRIPTION
**Pull request by feinsteinben**
Furthermore, replaced monkey-patches with (hopefully) more readable alternatives.

Breaking Changes
--------------------
1. The signature of the backward method has changed, and it is now:
```python
backward(inputs, grad_outputs)
```
where `inputs` and `grad_outputs` are tuples of tensors.
Previously, `inputs` and `grad_outputs` were both tensors.

### CODE REVIEW

**Comment by gwding** [comment]
@feinsteinben  Sorry for the delay in response. I reviewed the changes and it looks good to me. It is more readable than the original version and all the bpda tests passed.

One remaining question from me is that whether we can add additional tests to check the correctness of cases where `inputs` and `grad_outputs` are tuples of tensors?

**Comment by feinsteinben** [comment]
@gwding a test for this case was added.

File: tests/test_bpda.py [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 import pytest
 import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (gwding) [inline tests/test_bpda.py]: remove this line in order to pass flake8 tests https://travis-ci.org/BorealisAI/advertorch/builds/607893565?utm_source=github_status&utm_medium=notification

**Comment by gwding** [comment]
Looks good to me know, except for the 2 flake8 errors. suggestions made on how to fix those. will merge after these changes and travis-ci pass. Thanks again!

**Pull request merged**